source/tutorials/nix-language.md

```diff
@@ -0,0 +1,1679 @@
+
+### What do you need?
+
+- Familiarity with software development
+- Familiarity with Unix shell, to read command line examples <!-- TODO: link to yet-to-be instructions on "how to read command line examples" -->
+- A [Nix installation](./install-nix) to run the examples
+
+### How to run the examples?
+
+- A piece of Nix language code is a *Nix expression*.
+- Evaluating a Nix expression produces a *Nix value*.
+- The content of a *Nix file* (file extension `.nix`) is a Nix expression.
+
+:::{note}
+To *evaluate* means to transform an expression into a value according to the language rules.
+:::
+
+This guide contains many examples of Nix expressions.
+Each one is followed by the expected evaluation result.
+
+The following example is a Nix expression adding two numbers:
+
+```nix
+1 + 2
+```
+
+    3
+
+#### Interactive evaluation
+
+Use [`nix repl`][nix-repl] to evaluate Nix expressions interactively (by typing them on the command line):
+
+```console
+nix repl
+```
+
+    Welcome to Nix 2.5.1. Type :? for help.
+
+```console
+nix-repl> 1 + 2
+```
+
+    3
+
+:::{note}
+The Nix language by default uses lazy evaluation, and will only compute values when needed.
+
+Some examples show results of strict evaluation for clarity.
+If your output does not match the example, try prepending `:p` to the input expression.
+
+Example:
+
+```console
+nix-repl> { a.b.c = 1; }
+```
+
+    { a = { ... }; }
+
+
+```console
+nix-repl> :p { a.b.c = 1; }
+```
+
+    { a = { b = { c = 1; }; }; }
+
+:::
+
+[nix-repl]: https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-repl.html
+
+#### Evaluating Nix files
+
+Use [`nix-instantiate --eval`][nix-instantiate] to evaluate the expression in a Nix file.
+
+```console
+echo 1 + 2 > file.nix
+
+nix-instantiate --eval file.nix
+```
+
+    3
+
+<details><summary>Detailed explanation</summary>
+
+The first command writes `1 + 2` to a file `file.nix` in the current directory.
+The contents of `file.nix` are now `1 + 2`, which you can check with
+
+```console
+cat file.nix
+```
+
+    1 + 2
+
+The second command runs `nix-instantiate` with the `--eval` option on `file.nix`, which reads the file and evaluates the contained Nix expression.
+The resulting value is printed as output.
+
+`--eval` is required to evaluate the file and do nothing else.
+If `--eval` is omitted, `nix-instantiate` expects the expression in the given file to evaluate to a special value called *Derivation*, which we will come back to at the end of this guide in [](derivations).
+
+If you do not need `file.nix` any more, remove it with
+
+```console
+rm file.nix
+```
+
+</details>
+
+:::{note}
+`nix-instantiate --eval` will evaluate `default.nix` if no file name is specified.
+
+```console
+echo 1 + 2 > default.nix
+
+nix-instantiate --eval
+```
+
+    3
+:::
+
+:::{note}
+The Nix language by default uses lazy evaluation, and will only compute values when needed.
```

Comment:
Yes, intentional.
a) I expect people to scroll over this introductory boilerplate because they'd pick the `nix-repl` approach.
b) The invocation here is different and I think motivating it place is meaningful.